azkaban-common/src/main/java/azkaban/project/JdbcProjectImpl.java

```diff
@@ -802,6 +804,89 @@ public void uploadProjectProperties(final Project project,
     }
   }
 
+  @Override
+  public synchronized FlowRecommendation createFlowRecommendation(final int projectId, final String flowId)
+      throws ProjectManagerException {
+    logger.info("Creating flow recommendation " + flowId);
+    final String INSERT_FLOW_RECOMMENDATION =
+        "INSERT INTO project_flow_recommendations (project_id, flow_id, modified_time) values (?,?,?)";
+
+    final SQLTransaction<Integer> insertFlowRecommendation = transOperator ->
+        transOperator.update(INSERT_FLOW_RECOMMENDATION, projectId, flowId, System.currentTimeMillis());;
+
+    // Insert flow recommendation
+    try {
+      final int numRowsInserted = this.dbOperator.transaction(insertFlowRecommendation);
+      if (numRowsInserted == 0) {
+        throw new ProjectManagerException("No flow recommendations have been inserted.");
+      }
+    } catch (final SQLException ex) {
+      logger.error(INSERT_FLOW_RECOMMENDATION + " failed.", ex);
+      throw new ProjectManagerException("Insert flow recommendation " + flowId + " for existing "
+          + "project failed. ", ex);
+    }
+    return fetchFlowRecommendation(projectId, flowId);
```

Comment:
fetch also involve get DB. I think better to keep them (write and read) in a single transaction to avoid potential race condition.

Reply:
It should be fine
1. That write and read logic is copied from createNewProject function in the same file.
2. Every flow recommendation table read and write will be guarded by a lock in outer java classes.

Reply:
sounds good!